CMakeLists.txt

```diff
@@ -1102,173 +1073,151 @@ list(APPEND
-    ${out_S19_cmd}
-    )
-  list(APPEND
-    post_build_byproducts
-    ${KERNEL_S19_NAME}
-    ${out_S19_byprod}
-    )
+  get_property(elfconvert_formats TARGET bintools PROPERTY elfconvert_formats)
+  if(srec IN_LIST elfconvert_formats)
+    # Should we print a warning if case the tools does not support converting to s19 ?
+    list(APPEND
+      post_build_commands
+      COMMAND $<TARGET_PROPERTY:bintools,elfconvert_command>
+              $<TARGET_PROPERTY:bintools,elfconvert_flag>
+              ${GAP_FILL}
+              $<TARGET_PROPERTY:bintools,elfconvert_flag_outtarget>srec
+              $<TARGET_PROPERTY:bintools,elfconvert_flag_srec_len>1
+              $<TARGET_PROPERTY:bintools,elfconvert_flag_infile>${KERNEL_ELF_NAME}
+              $<TARGET_PROPERTY:bintools,elfconvert_flag_outfile>${KERNEL_S19_NAME}
+      )
+    list(APPEND
+      post_build_byproducts
+      ${KERNEL_S19_NAME}
+      # ${out_S19_byprod} # Is this needed ?
+
+      )
+  endif()
 endif()
 
 if(CONFIG_OUTPUT_DISASSEMBLY)
-  set(out_disassembly_cmd    "")
-  set(out_disassembly_byprod "")
-  if(CONFIG_OUTPUT_DISASSEMBLE_ALL)
-    set(disassembly_type DISASSEMBLE_ALL)
+if(CONFIG_OUTPUT_DISASSEMBLE_ALL)
+    set(disassembly_type "$<TARGET_PROPERTY:bintools,disassembly_flag_all>")
   else()
-    set(disassembly_type DISASSEMBLE_SOURCE)
+    set(disassembly_type "$<TARGET_PROPERTY:bintools,disassembly_flag_inline_source>")
   endif()
-  bintools_objdump(
-    RESULT_CMD_LIST    out_disassembly_cmd
-    RESULT_BYPROD_LIST out_disassembly_byprod
-    ${disassembly_type}
-    FILE_INPUT         ${KERNEL_ELF_NAME}
-    FILE_OUTPUT        ${KERNEL_LST_NAME}
-    )
   list(APPEND
     post_build_commands
-    ${out_disassembly_cmd}
+    COMMAND $<TARGET_PROPERTY:bintools,disassembly_command>
+            $<TARGET_PROPERTY:bintools,disassembly_flag>
+            ${disassembly_type}
+            $<TARGET_PROPERTY:bintools,disassembly_flag_infile>${KERNEL_ELF_NAME}
+            $<TARGET_PROPERTY:bintools,disassembly_flag_outfile>${KERNEL_LST_NAME}
     )
   list(APPEND
     post_build_byproducts
     ${KERNEL_LST_NAME}
-    ${out_disassembly_byprod}
+#    ${out_disassembly_byprod} # Needed ??
     )
 endif()
 
 if(CONFIG_OUTPUT_STAT)
-  set(out_stat_cmd    "")
-  set(out_stat_byprod "")
-  bintools_readelf(
-    RESULT_CMD_LIST    out_stat_cmd
-    RESULT_BYPROD_LIST out_stat_byprod
-    HEADERS
-    FILE_INPUT         ${KERNEL_ELF_NAME}
-    FILE_OUTPUT        ${KERNEL_STAT_NAME}
-    )
+#  zephyr_post_build(TOOLS bintools COMMAND readelf FLAGS headers INFILE file OUTFILE outfile)
   list(APPEND
     post_build_commands
-    ${out_stat_cmd}
+    COMMAND $<TARGET_PROPERTY:bintools,readelf_command>
+            $<TARGET_PROPERTY:bintools,readelf_flag>
+            $<TARGET_PROPERTY:bintools,readelf_flag_headers>
+            $<TARGET_PROPERTY:bintools,readelf_flag_infile> ${KERNEL_ELF_NAME}
+            $<TARGET_PROPERTY:bintools,readelf_flag_outfile> ${KERNEL_STAT_NAME}
```

Comment:
> bintools functions, and placing the command construction into the top level CMakeLists.txt file, we are essentially locking the format of the command construction, thus removing the option for the toolchain bintools to completely redefine the way the command is constructed. 

This is valid concern.
I tried to keep the names and arguments as open as possible while of course having GNU bintools as reference cause for time being there are no other reference points.
That's why there are:
- `<command>_flag` for all, where any flag can be placed.
- `<command>_infile`
- `<command>_outfile`

The names are `elfconvert`, `disassembly` as this is how they are used, and not called `objcopy` and `objdump`.
And the nice part in this approach is that we can easily add additional flags if tool Y needs those, as they can be empty for GNU bintools.
So far I also took into consideration the arguments that were used for the macros, cause if the macro were called, for example:
```
bintools_objcopy(
    RESULT_CMD_LIST obj_copy_cmd
    TARGET_INPUT    ${OUTPUT_FORMAT}
    TARGET_OUTPUT   "binary"
    SECTION_ONLY    ".intList"
    FILE_INPUT      $<TARGET_FILE:${ZEPHYR_PREBUILT_EXECUTABLE}>
    FILE_OUTPUT     "isrList.bin"
    )
```
then I made the assumption that `TARGET_INPUT`, `TARGET_OUTPUT`, `SECTION_ONLY`, `FILE_INPUT`, `FILE_OUTPUT` are all required. But of course, the macro could decide to use `SECRET_INTERNAL_SETTING` also.
Supporting such use is what I hoped to ensure with `<command>_flag` property.

We can extend with a `<command>_raw` or `<command>_custom` that could be used to overrule all settings.
I do see the benefit in macros, but it also bears the risk that they may be misused, as to suddenly specify `compiler` flags in an elfconvert command. If that is needed, then we might want to have such possibility exposed in compiler flags instead.

Only risk would of course be if argument order matters, that is `--arg_A` must be placed before `--arg_B`.
But I wanted to wait doing so, unless there was a specific need.

> The only exception to this seems to be the memusage command, which is definable by a property.

Because that was the only place I could see it was already needed.
We can allow similar on all commands.

> which can then be changed as needed by the toolchain bintools.

If by this you mean that first time it is called, then it uses args `-A -B -C` and second time then `-D -E -F`, then that is gonna be a little tricky, cause the command itself is constructed using a generator expression.

> Is there any way we could keep the flexibility of the functions

Are you looking for an ability to hook in a completely custom command to `post_build_commands` ?
It's a fine balance on making things easy and consistent, while still allowing for special use-cases.

Reply:
I completely agree in that this is a fine balance between flexibility, consistency and ease of use, and the uses for the custom command construction is probably very limited, thus meaning that ease of use and consistency weigh highest.

If it is not going to be possible to keep the function based bintools abstraction, due to consistency, having a `<command>_raw` or similar, seems like a good option. This would though need to one command raw option per configure section, such as one for `CONFIG_OUTPUT_DISASSEMBLY as bintool_output_disassembly_command` and another for `CONFIG_OUTPUT_STAT as bintool_output_stat_command` and so forth.
The above would also allow for argument reordering, where and if needed.

> If by this you mean that first time it is called, then it uses args -A -B -C and second time then -D -E -F
I did not, rather what I described above as a customizable command for each configure flag.

> Are you looking for an ability to hook in a completely custom command to post_build_commands ?
No, only as above, per configure flag command. Perhaps though, that would also be a good feature to have, but I currently have no particular use case in mind.

Reply:
> If it is not going to be possible to keep the function based bintools abstraction, due to consistency, having a `<command>_raw` or similar, seems like a good option. This would though need to one command raw option per configure section, such as one for `CONFIG_OUTPUT_DISASSEMBLY as bintool_output_disassembly_command` and another for `CONFIG_OUTPUT_STAT as bintool_output_stat_command` and so forth.
> The above would also allow for argument reordering, where and if needed.

We actually have this in place today with this PR, it just need additional documentation.
What a user can do is simply to create a custom CMake script that wil invoke whatever command he likes.
```
set_property(TARGET bintools PROPERTY readelf_command ${CMAKE_COMMAND})

set_property(TARGET bintools PROPERTY readelf_flag "")
set_property(TARGET bintools PROPERTY readelf_flag_headers "")

set_property(TARGET bintools PROPERTY readelf_flag_infile "-DINFILE=")
set_property(TARGET bintools PROPERTY readelf_flag_outfile "-DOUTFILE=")

set_property(TARGET bintools PROPERTY readelf_flag_final -P my_custom_readelf.cmake)
```

and then user can place anything inside the `my_custom_readelf.cmake` script.
And order arguments can be reordered as desired.

Of course, the best is if we can have such a generic pattern that 99% of use cases can be covered without the need of a custom script, but it's good to have the possibility from start.

Only thing to do, is to allow a `<command>_flag_final` on all commands.

Reply:
@daor-oti Done, see suggestion here: https://github.com/zephyrproject-rtos/zephyr/pull/24851/commits/b921627c9169edab985899dfc098180a8bda458e

And description:
```
# The bintools properties are made generic so that implementing support for an
# additional native tool should be as easy as possible.
# However, there might be tools and/or use cases where the current property
# scheme does not cover the exact needs. For those use-cases it is possible
# to implement the call to a native tool inside a CMake script.
# For example, to call a custom script for elfconvert command, one can specify:
#   set_property(TARGET bintools PROPERTY elfconvert_command ${CMAKE_COMMAND})
#   set_property(TARGET bintools PROPERTY elfconvert_flag "")
#   set_property(TARGET bintools PROPERTY elfconvert_flag_final     -P custom_elfconvert.cmake)
#   set_property(TARGET bintools PROPERTY elfconvert_flag_strip_all "-DSTRIP_ALL=True")
#   set_property(TARGET bintools PROPERTY elfconvert_flag_infile    "-DINFILE=")
#   set_property(TARGET bintools PROPERTY elfconvert_flag_outfile   "-DOUT_FILE=")
```

Reply:
@tejlmand Thanks, this looks good.